aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs

```diff
@@ -0,0 +1,648 @@
+                match subtraction(self.value, value) {
+                    Some(result) => {
+                        self.value = result;
+                        Ok(())
+                    }
+                    None => Err(abort_error(
+                        format!(
+                            "aggregator's value underflowed when subtracting {} from {}",
+                            value, self.value
+                        ),
+                        EAGGREGATOR_UNDERFLOW,
+                    )),
+                }
+            }
+            // Since `sub` is a barrier, we never encounter delta during
+            // subtraction. This will change in future.
+            // TODO: implement this properly once `sub` stops being a barrier.
+            AggregatorState::PositiveDelta => {
+                unreachable!("subtraction always materializes the value")
+            }
+        }
+    }
+
+    /// Materializes the value of the aggregator, changing its state.
+    fn materialize(
+        &mut self,
+        context: &NativeAggregatorContext,
+        handle: &TableHandle,
+    ) -> PartialVMResult<()> {
+        // If aggregator has already been materialized, return immediately.
+        if self.state == AggregatorState::Data {
+            return Ok(());
+        }
+
+        // Otherwise, we have a delta and have to go to storage.
+        let key_bytes = serialize(&self.key);
+        match context.resolver.resolve_table_entry(handle, &key_bytes) {
+            Err(_) => Err(extension_error("error resolving aggregator's value")),
+            Ok(maybe_bytes) => {
+                match maybe_bytes {
+                    Some(bytes) => {
+                        let base = deserialize(&bytes);
+
+                        // The only remaining case is PositiveDelta. But assert just in case.
+                        assert!(self.state == AggregatorState::PositiveDelta);
+
+                        // At this point, we fetched the "true" value of the
+                        // aggregator from the previously executed transaction.
+                        match addition(base, self.value, self.limit) {
+                            Some(result) => {
+                                self.value = result;
+                                self.state = AggregatorState::Data;
+                                Ok(())
+                            }
+                            None => Err(abort_error(
+                                "aggregator's value overflowed when materializing the delta",
+                                EAGGREGATOR_OVERFLOW,
+                            )),
+                        }
+                    }
+                    None => Err(extension_error("aggregator's value not found")),
+                }
+            }
+        }
+    }
+}
+
+/// Aggregator table - a top-level collection storing aggregators. It holds an
+/// inner table data structure to actually store the aggregators.
+struct AggregatorTable {
```

Comment:
Maybe let's call this registry since the underlying data-structure is also a table to avoid confusion.
Here somewhere in comments let's say this is per context.